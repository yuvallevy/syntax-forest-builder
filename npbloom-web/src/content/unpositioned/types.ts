import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;

export type PlotCoordsOffset = {
  dPlotX: DPlotX;
  dPlotY: DPlotY;
};

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: PlotCoordsOffset;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: Id[];
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: StringSlice;
  triangle: boolean;
};

export type UnpositionedFormerlyBranchingNode = NodeCommon & WithOffsetInTree & {
  formerDescendants: IdMap<UnpositionedNode>;
};

export type UnpositionedFormerlyTerminalNode = NodeCommon & WithOffsetInTree & {
  formerSlice: StringSlice;
  formerlyTriangle: boolean;
};

export type UnpositionedPlainStrandedNode = NodeCommon & WithOffsetInTree;

export type UnpositionedStrandedNode = UnpositionedFormerlyBranchingNode | UnpositionedFormerlyTerminalNode |
  UnpositionedPlainStrandedNode;

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode | UnpositionedStrandedNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export const isTerminal = (node: UnpositionedNode): node is UnpositionedTerminalNode =>
  'slice' in node;

export const isFormerlyBranching = (node: UnpositionedNode): node is UnpositionedFormerlyBranchingNode =>
  'formerDescendants' in node;

export const isFormerlyTerminal = (node: UnpositionedNode): node is UnpositionedFormerlyTerminalNode =>
  'formerSlice' in node;
