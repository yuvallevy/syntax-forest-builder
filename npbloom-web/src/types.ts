import { ContentChange, ContentState, UndoRedoHistory } from 'npbloom-core';

export type Id = string;
export type IdMap<T> = Record<Id, T>
export type Sentence = string;
export type PlotIndex = number;
export type NodeLabel = string;
export type UndoableContentState = UndoRedoHistory<ContentState, ContentChange>;
