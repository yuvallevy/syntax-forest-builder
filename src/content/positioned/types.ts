import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type Width = number;

export type PlotCoords = {
  plotX: PlotX;
  plotY: PlotY;
};

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: Id[];
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  slice: StringSlice;
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedStrandedNode = NodeCommon & WithPositionInTree;

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode | PositionedStrandedNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: PlotCoords;
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
