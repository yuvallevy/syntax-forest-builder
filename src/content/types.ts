export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;

export type StringSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type NodeIndicatorInPlot = {
  treeId: Id;
  nodeId: Id;
};

export type TreeCommon = {
  sentence: Sentence;
}

export type NodeCommon = {
  label: NodeLabel;
};
