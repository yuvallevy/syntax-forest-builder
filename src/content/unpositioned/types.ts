import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;

export type PlotCoordsOffset = {
  dPlotX: DPlotX;
  dPlotY: DPlotY;
};

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: PlotCoordsOffset;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: Id[];
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: StringSlice;
  triangle: boolean;
};

export type UnpositionedStrandedNode = NodeCommon & WithOffsetInTree & {
  formerDescendants?: IdMap<UnpositionedNode>;
  formerSlice?: StringSlice;
  formerlyTriangle?: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode | UnpositionedStrandedNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export const isTerminal = (node: UnpositionedNode): node is UnpositionedTerminalNode =>
  'slice' in node;
