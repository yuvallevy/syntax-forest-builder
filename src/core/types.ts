export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;
type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;
type Width = number;

export type NodeSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type UnpositionedForest = { plots: IdMap<UnpositionedPlot>; };

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

type TreeCommon = {
  sentence: Sentence;
}

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: { dPlotX: DPlotX; dPlotY: DPlotY; };
};

export type NodeCommon = {
  label: NodeLabel;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: IdMap<UnpositionedNode>;
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: NodeSlice;
  triangle: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: IdMap<PositionedNode>;
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: {
    plotX: PlotX;
    plotY: PlotY;
  };
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
