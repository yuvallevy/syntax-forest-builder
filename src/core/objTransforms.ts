export const transformValues = <K extends string, V, R>(record: Record<K, V>, valueTransform: (value: V) => R): Record<K, R> =>
  (Object.entries(record) as [K, V][]).reduce((transformedRecord, [key, value]) => ({
    ...transformedRecord,
    [key]: valueTransform(value),
  }), {} as Record<K, R>);

export const mapValues = <K extends string, V, R>(record: Record<K, V>, valueTransform: (value: V) => R): R[] =>
  (Object.values(record) as V[]).map(valueTransform);

export const mapEntries = <K extends string, V, R>(record: Record<K, V>, entryTransform: (entry: [K, V]) => R): R[] =>
  (Object.entries(record) as [K, V][]).map(entryTransform);

export const without = <T>(array: T[], elementToRemove: T): T[] =>
  array.reduce((accum, element) => element === elementToRemove ? accum : [...accum, element], [] as T[]);

export const union = <T>(array1: T[], array2: T[]): T[] =>
  array2.reduce((accum, element) => array1.includes(element) ? accum : [...accum, element], array1);

export const omitKey = <K extends string, V>(record: Record<K, V>, keyToOmit: K): Record<K, V> =>
  record.hasOwnProperty(keyToOmit)
    ? (Object.entries(record) as [K, V][]).reduce((transformedRecord, [key, value]) => key === keyToOmit ? transformedRecord : {
      ...transformedRecord,
      [key]: value,
    }, {} as Record<K, V>)
    : record;

export const isEmpty = (objOrArray: any): boolean => (objOrArray instanceof Array ? objOrArray : Object.keys(objOrArray)).length === 0;
