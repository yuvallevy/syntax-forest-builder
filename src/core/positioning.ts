import { mapValues, transformValues } from './objTransforms';
import { PositionedNode, PositionedPlot, PositionedTree, Sentence, UnpositionedNode, UnpositionedPlot, UnpositionedTree } from './types';

const DEFAULT_TERMINAL_NODE_Y = -2;
const DEFAULT_TRIANGLE_NODE_Y = -20;
const DEFAULT_NODE_LEVEL_DIFF = -40;

type StrWidthFunc = (str: string) => number;

/**
 * Returns a node and its descendants, if any, with positions.
 */
const applyNodePosition = (strWidthFunc: StrWidthFunc) => (sentence: Sentence) => (node: UnpositionedNode | PositionedNode): PositionedNode => {
  if ('position' in node) {
    // This node already has a position
    return node;
  }

  if ('slice' in node) {
    // This is a terminal node
    const [sliceStart, sliceEnd] = node.slice;
    const widthBeforeSlice = strWidthFunc(sentence.slice(0, sliceStart));
    const sliceWidth = strWidthFunc(sentence.slice(sliceStart, sliceEnd));
    return {
      ...node,
      position: {
        treeX: widthBeforeSlice + (sliceWidth / 2) + node.offset.dTreeX,
        treeY: (node.triangle ? DEFAULT_TRIANGLE_NODE_Y : DEFAULT_TERMINAL_NODE_Y) + node.offset.dTreeY,
      },
      triangle: node.triangle ? {
        treeX1: widthBeforeSlice,
        treeX2: widthBeforeSlice + sliceWidth,
      } : undefined,
    };
  }

  // This is a branching node
  const positionedChildNodes = transformValues(node.children, applyNodePosition(strWidthFunc)(sentence));
  const childXs = mapValues(positionedChildNodes, node => node.position.treeX);
  const childYs = mapValues(positionedChildNodes, node => node.position.treeY);
  return {
    ...node,
    children: positionedChildNodes,
    position: {
      treeX: childXs.reduce((accum, cur) => accum + cur, 0) / childXs.length + node.offset.dTreeX,
      treeY: Math.min(...childYs) + DEFAULT_NODE_LEVEL_DIFF + node.offset.dTreeY,
    },
  };
}

/**
 * Returns a copy of the given tree with positions for all nodes.
 */
export const applyNodePositionsToTree = (strWidthFunc: StrWidthFunc) => (tree: UnpositionedTree): PositionedTree => ({
  ...tree,
  nodes: transformValues(tree.nodes, applyNodePosition(strWidthFunc)(tree.sentence)),
  position: {
    plotX: tree.offset.dPlotX,
    plotY: tree.offset.dPlotY,
  },
  width: strWidthFunc(tree.sentence),
});

/**
 * Returns a copy of the given plot with positions for all trees and nodes.
 */
export const applyNodePositionsToPlot = (strWidthFunc: StrWidthFunc) => (plot: UnpositionedPlot): PositionedPlot => ({
  ...plot,
  trees: transformValues(plot.trees, applyNodePositionsToTree(strWidthFunc)),
});
