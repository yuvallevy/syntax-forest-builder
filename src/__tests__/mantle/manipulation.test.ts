import { describe, expect, it } from 'vitest';
import { UnpositionedTree } from '../../core/types';
import {
  deleteNodesInTree,
  getNodeIdsAssignedToSlice,
  getParentNodeIdsInTree,
  insertNodeIntoTree,
  transformAllNodesInTree,
  transformNodeInTree,
  transformNodesInTree,
} from '../../mantle/manipulation';

describe('tree manipulation', () => {
  const tree: UnpositionedTree = {
    sentence: 'Noun verbed.',
    nodes: {
      'top': { label: 'S', offset: { dTreeX: 0, dTreeY: 5 }, children: ['branch1', 'term2'] },
      'branch1': { label: 'NP', offset: { dTreeX: 0, dTreeY: 0 }, children: ['term1'] },
      'term1': { label: 'N', offset: { dTreeX: 0, dTreeY: 0 }, slice: [0, 4], triangle: false },
      'term2': { label: 'VP', offset: { dTreeX: 0, dTreeY: 0 }, slice: [5, 11], triangle: false },
    },
    offset: { dPlotX: 0, dPlotY: 0 },
  };

  const treeWithoutTopLevelTerminalNode: UnpositionedTree = {
    ...tree,
    nodes: {
      'top': { label: 'S', offset: { dTreeX: 0, dTreeY: 5 }, children: ['branch1'] },
      'branch1': { label: 'NP', offset: { dTreeX: 0, dTreeY: 0 }, children: ['term1'] },
      'term1': { label: 'N', offset: { dTreeX: 0, dTreeY: 0 }, slice: [0, 4], triangle: false },
    },
  };

  const treeWithoutTopLevelBranchingNode: UnpositionedTree = {
    ...tree,
    nodes: {
      'branch1': { label: 'NP', offset: { dTreeX: 0, dTreeY: 0 }, children: ['term1'] },
      'term1': { label: 'N', offset: { dTreeX: 0, dTreeY: 0 }, slice: [0, 4], triangle: false },
      'term2': { label: 'VP', offset: { dTreeX: 0, dTreeY: 0 }, slice: [5, 11], triangle: false },
    },
  };

  const changeLabel = transformNodeInTree(node => ({ ...node, label: 'test' }));

  it('retrieves the parent ID of a single node', () => {
    expect(getParentNodeIdsInTree(['term1'])(tree)).toStrictEqual(['branch1']);
  });

  it('retrieves the parent ID of two sibling nodes', () => {
    expect(getParentNodeIdsInTree(['branch1', 'term2'])(tree)).toStrictEqual(['top']);
  });

  it('retrieve the parent IDs of two non-sibling nodes', () => {
    expect(getParentNodeIdsInTree(['branch1', 'term1'])(tree)).toStrictEqual(['top', 'branch1']);
  });

  it('retrieves no parent IDs for a top-level node', () => {
    expect(getParentNodeIdsInTree(['top'])(tree)).toStrictEqual([]);
  });

  it('retrieves only one parent ID for two non-sibling nodes of which one is a top-level node', () => {
    expect(getParentNodeIdsInTree(['top', 'term1'])(tree)).toStrictEqual(['branch1']);
  });

  it('retrieves the terminal ID associated with a slice when there is only one', () => {
    expect(getNodeIdsAssignedToSlice([5, 11])(tree)).toStrictEqual(['term2']);
  });

  it('retrieves the terminal ID associated with a slice when there is only one and it covers more than the slice', () => {
    expect(getNodeIdsAssignedToSlice([6, 10])(tree)).toStrictEqual(['term2']);
  });

  it('retrieves all terminal IDs associated with a slice when there are multiple', () => {
    expect(getNodeIdsAssignedToSlice([3, 9])(tree)).toStrictEqual(['term1', 'term2']);
  });

  it('retrieves no terminal IDs associated with a slice when there are none', () => {
    expect(getNodeIdsAssignedToSlice([5, 11])(treeWithoutTopLevelTerminalNode)).toStrictEqual([]);
  });

  it('inserts a top-level terminal node', () => {
    expect(insertNodeIntoTree({
      label: 'V',
      targetSlice: [5, 11],
    })('new')(treeWithoutTopLevelTerminalNode)).toMatchSnapshot();
  });

  it('inserts a top-level branching node with one child', () => {
    expect(insertNodeIntoTree({
      label: 'VP',
      targetChildIds: ['term2'],
    })('new')(treeWithoutTopLevelBranchingNode)).toMatchSnapshot();
  });

  it('inserts a top-level branching node with two children', () => {
    expect(insertNodeIntoTree({
      label: 'S',
      targetChildIds: ['branch1', 'term2'],
    })('new')(treeWithoutTopLevelBranchingNode)).toMatchSnapshot();
  });

  it('inserts a branching node down the tree with a terminal child', () => {
    expect(insertNodeIntoTree({
      label: 'N\'',
      targetParentId: 'branch1',
      targetChildIds: ['term1'],
    })('new')(tree)).toMatchSnapshot();
  });

  it('inserts a branching node down the tree with a branching child', () => {
    expect(insertNodeIntoTree({
      label: '?',
      targetParentId: 'top',
      targetChildIds: ['branch1'],
    })('new')(tree)).toMatchSnapshot();
  });

  it('inserts a branching node down the tree with two children', () => {
    expect(insertNodeIntoTree({
      label: '?',
      targetParentId: 'top',
      targetChildIds: ['branch1', 'term2'],
    })('new')(tree)).toMatchSnapshot();
  });

  it('changes the label of a node', () => {
    expect(changeLabel('branch1')(tree).nodes).toMatchSnapshot();
  });

  it('changes trianglehood of two nodes at the same time', () => {
    expect(transformNodesInTree(node => ({ ...node, triangle: true }))(['term1', 'term2'])(tree).nodes)
      .toMatchSnapshot();
  });

  it('changes all nodes in the tree', () => {
    expect(transformAllNodesInTree(node => ({ ...node, label: 'test' }))(tree).nodes).toMatchSnapshot();
  });

  it('deletes a top-level node', () => {
    expect(deleteNodesInTree(['top'])(tree).nodes).toMatchSnapshot();
  });

  it('deletes a branching node down the tree', () => {
    expect(deleteNodesInTree(['branch1'])(tree).nodes).toMatchSnapshot();
  });

  it('deletes a terminal node down the tree', () => {
    expect(deleteNodesInTree(['term2'])(tree).nodes).toMatchSnapshot();
  });

  it('deletes two nodes with direct parent-child relationship at once', () => {
    expect(deleteNodesInTree(['top', 'term2'])(tree).nodes).toMatchSnapshot();
  });

  it('deletes two nodes with ancestor-descendant relationship at once', () => {
    expect(deleteNodesInTree(['top', 'term1'])(tree).nodes).toMatchSnapshot();
  });

  it('deletes two nodes with no ancestor-descendant relationship at once', () => {
    expect(deleteNodesInTree(['term1', 'term2'])(tree).nodes).toMatchSnapshot();
  });

  it('makes a branching node stranded when its children are gone', () => {
    expect(deleteNodesInTree(['branch1', 'term2'])(tree).nodes).toMatchSnapshot();
  });

  describe('referential equality of nodes after changes', () => {
    it('keeps unaffected nodes', () => {
      const treeAfterChange = changeLabel('branch1')(tree);
      expect(tree.nodes['term2']).toBe(treeAfterChange.nodes['term2']);
    });
  
    it('keeps children of deleted nodes', () => {
      const treeAfterChange = deleteNodesInTree(['top'])(tree);
      expect(tree.nodes['branch1']).toBe(treeAfterChange.nodes['branch1']);
      expect(tree.nodes['term2']).toBe(treeAfterChange.nodes['term2']);
    });
  
    it('changes changed nodes', () => {
      const treeAfterChange = changeLabel('branch1')(tree);
      expect(tree.nodes['branch1']).not.toBe(treeAfterChange.nodes['branch1']);
    });
  
    it('changes parents of deleted nodes', () => {
      const treeAfterChange = deleteNodesInTree(['branch1'])(tree);
      expect(tree.nodes['top']).not.toBe(treeAfterChange.nodes['top']);
    });
  });
});
