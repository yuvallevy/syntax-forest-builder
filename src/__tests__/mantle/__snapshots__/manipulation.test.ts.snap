// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tree manipulation changes all nodes in the tree 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
    ],
    "label": "test",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "test",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      4,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "test",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      5,
      11,
    ],
    "triangle": false,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term2",
    ],
    "label": "test",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`tree manipulation changes the label of a node 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
    ],
    "label": "test",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      4,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      5,
      11,
    ],
    "triangle": false,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term2",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`tree manipulation deletes a branching node down the tree 1`] = `
Object {
  "term1": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      4,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      5,
      11,
    ],
    "triangle": false,
  },
  "top": Object {
    "children": Array [
      "term2",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`tree manipulation deletes a terminal node down the tree 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      4,
    ],
    "triangle": false,
  },
  "top": Object {
    "children": Array [
      "branch1",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`tree manipulation deletes a top-level node 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      4,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      5,
      11,
    ],
    "triangle": false,
  },
}
`;

exports[`tree manipulation deletes two nodes with ancestor-descendant relationship at once 1`] = `
Object {
  "branch1": Object {
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term2": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      5,
      11,
    ],
    "triangle": false,
  },
}
`;

exports[`tree manipulation deletes two nodes with direct parent-child relationship at once 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      4,
    ],
    "triangle": false,
  },
}
`;

exports[`tree manipulation deletes two nodes with no ancestor-descendant relationship at once 1`] = `
Object {
  "branch1": Object {
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "top": Object {
    "children": Array [
      "branch1",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`tree manipulation inserts a branching node down the tree with a branching child 1`] = `
Object {
  "nodes": Object {
    "branch1": Object {
      "children": Array [
        "term1",
      ],
      "label": "NP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "new": Object {
      "children": Array [
        "branch1",
      ],
      "label": "?",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "term1": Object {
      "label": "N",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        0,
        4,
      ],
      "triangle": false,
    },
    "term2": Object {
      "label": "VP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        5,
        11,
      ],
      "triangle": false,
    },
    "top": Object {
      "children": Array [
        "term2",
        "new",
      ],
      "label": "S",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 5,
      },
    },
  },
  "offset": Object {
    "dPlotX": 0,
    "dPlotY": 0,
  },
  "sentence": "Noun verbed.",
}
`;

exports[`tree manipulation inserts a branching node down the tree with a terminal child 1`] = `
Object {
  "nodes": Object {
    "branch1": Object {
      "children": Array [
        "new",
      ],
      "label": "NP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "new": Object {
      "children": Array [
        "term1",
      ],
      "label": "N'",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "term1": Object {
      "label": "N",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        0,
        4,
      ],
      "triangle": false,
    },
    "term2": Object {
      "label": "VP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        5,
        11,
      ],
      "triangle": false,
    },
    "top": Object {
      "children": Array [
        "branch1",
        "term2",
      ],
      "label": "S",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 5,
      },
    },
  },
  "offset": Object {
    "dPlotX": 0,
    "dPlotY": 0,
  },
  "sentence": "Noun verbed.",
}
`;

exports[`tree manipulation inserts a branching node down the tree with two children 1`] = `
Object {
  "nodes": Object {
    "branch1": Object {
      "children": Array [
        "term1",
      ],
      "label": "NP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "new": Object {
      "children": Array [
        "branch1",
        "term2",
      ],
      "label": "?",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "term1": Object {
      "label": "N",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        0,
        4,
      ],
      "triangle": false,
    },
    "term2": Object {
      "label": "VP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        5,
        11,
      ],
      "triangle": false,
    },
    "top": Object {
      "children": Array [
        "new",
      ],
      "label": "S",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 5,
      },
    },
  },
  "offset": Object {
    "dPlotX": 0,
    "dPlotY": 0,
  },
  "sentence": "Noun verbed.",
}
`;

exports[`tree manipulation inserts a top-level branching node with one child 1`] = `
Object {
  "nodes": Object {
    "branch1": Object {
      "children": Array [
        "term1",
      ],
      "label": "NP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "new": Object {
      "children": Array [
        "term2",
      ],
      "label": "VP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "term1": Object {
      "label": "N",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        0,
        4,
      ],
      "triangle": false,
    },
    "term2": Object {
      "label": "VP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        5,
        11,
      ],
      "triangle": false,
    },
  },
  "offset": Object {
    "dPlotX": 0,
    "dPlotY": 0,
  },
  "sentence": "Noun verbed.",
}
`;

exports[`tree manipulation inserts a top-level branching node with two children 1`] = `
Object {
  "nodes": Object {
    "branch1": Object {
      "children": Array [
        "term1",
      ],
      "label": "NP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "new": Object {
      "children": Array [
        "branch1",
        "term2",
      ],
      "label": "S",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "term1": Object {
      "label": "N",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        0,
        4,
      ],
      "triangle": false,
    },
    "term2": Object {
      "label": "VP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        5,
        11,
      ],
      "triangle": false,
    },
  },
  "offset": Object {
    "dPlotX": 0,
    "dPlotY": 0,
  },
  "sentence": "Noun verbed.",
}
`;

exports[`tree manipulation inserts a top-level terminal node 1`] = `
Object {
  "nodes": Object {
    "branch1": Object {
      "children": Array [
        "term1",
      ],
      "label": "NP",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
    },
    "new": Object {
      "label": "V",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        5,
        11,
      ],
      "triangle": false,
    },
    "term1": Object {
      "label": "N",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 0,
      },
      "slice": Array [
        0,
        4,
      ],
      "triangle": false,
    },
    "top": Object {
      "children": Array [
        "branch1",
        "term2",
      ],
      "label": "S",
      "offset": Object {
        "dTreeX": 0,
        "dTreeY": 5,
      },
    },
  },
  "offset": Object {
    "dPlotX": 0,
    "dPlotY": 0,
  },
  "sentence": "Noun verbed.",
}
`;

exports[`tree manipulation makes a branching node stranded when its children are gone 1`] = `
Object {
  "term1": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      4,
    ],
    "triangle": false,
  },
  "top": Object {
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;
